src/parse/validator.rs

```diff
@@ -250,131 +257,136 @@ impl<'help, 'app, 'parser> Validator<'help, 'app, 'parser> {
-                    name
-                );
-                should_err = ma.occurs > 0;
-            }
-            if should_err {
-                return self.build_conflict_err(name, matcher);
-            }
-        }
-        Ok(())
+                    .find(|g| !g.multiple && g.id == *name)
+                {
+                    let conf_with_self = || {
+                        self.p
+                            .app
+                            .unroll_args_in_group(&g.id)
+                            .iter()
+                            .filter(|&a| matcher.contains(a))
+                            .count()
+                            > 1
+                    };
+                    let conf_with_arg = || g.conflicts.iter().any(|x| matcher.contains(x));
+                    let arg_conf_with_gr = || {
+                        matcher
+                            .arg_names()
+                            .filter_map(|x| self.p.app.find(x))
+                            .any(|x| x.blacklist.iter().any(|c| *c == g.id))
+                    };
+                    conf_with_self() || conf_with_arg() || arg_conf_with_gr()
+                } else if let Some(ma) = matcher.get(name) {
+                    debug!(
+                        "Validator::validate_conflicts:iter:{:?}: matcher contains it...",
+                        name
+                    );
+                    ma.occurs > 0
+                } else {
+                    false
+                }
+            })
+            // Throw an error for the first conflict found.
+            .try_for_each(|odd| self.build_conflict_err(odd, matcher))
     }
 
     fn validate_exclusive(&mut self, matcher: &mut ArgMatcher) -> ClapResult<()> {
         debug!("Validator::validate_exclusive");
         let args_count = matcher.arg_names().count();
-        for name in matcher.arg_names() {
-            debug!("Validator::validate_exclusive:iter:{:?}", name);
-            if let Some(arg) = self.p.app.find(name) {
-                if arg.exclusive && args_count > 1 {
-                    let c_with: Option<String> = None;
-                    return Err(Error::argument_conflict(
-                        arg,
-                        c_with,
-                        Usage::new(self.p).create_usage_with_title(&[]),
-                        self.p.app.color(),
-                    ));
-                }
-            }
-        }
-        Ok(())
+        matcher
+            .arg_names()
+            .filter_map(|name| {
+                debug!("Validator::validate_exclusive:iter:{:?}", name);
+                self.p
+                    .app
+                    .find(name)
+                    // Find `arg`s which are exclusive but also appear with other args.
+                    .filter(|&arg| arg.exclusive && args_count > 1)
+            })
+            // Throw an error for the first conflict found.
+            .try_for_each(|arg| {
```

Comment:
I have a theoretical question here. The old code you removed fails faster and thus improves performance. Does the new code you wrote do the same?

Reply:
Firstly, this change will **not affect** short-circuiting (aka. "early return"): iterators are naturally lazy, and short-circuiting is assured by the way in which [`try_fold`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold) is implemented.

What's more, this change enhances **composition** of operations, allowing us to add more conditions in a much more easier way in the future.